File: java/ll.java
// Code to implement singly linked list
// Author: Aditya Karia


public class ll
{
	public class node
	{
		int value;
		node next;

		node(int data)
		{
			this.value = data;
			this.next = null;
		}
	}

	node head = new node(0);

	void insert_end(int x)
	{
		node temp = new node(x);
		node cur = head;
		
		while (cur.next != null)
		{
			cur = cur.next;
		}

		cur.next = temp;
	}

	public void insert_at_index(int index, int x)
	{
		node temp = new node(x);

		node cur = head;
		
		for (int i = 0; i < index && cur.next != null; i++)
		{
			cur = cur.next;
		}

		temp.next = cur.next;
		cur.next = temp;
	}

	public void printlist()
	{
		System.out.println("List is:");
		node cur = head;

		while (cur.next != null)
		{
			cur = cur.next;
			System.out.println(cur.value);
		}
		System.out.println("");
	}

	public static void main(String[] args)
	{
		ll mylist = new ll();
		
		int array[] = {1,2,3,4,5,6,7,8,9,69};

		for (int i: array)
		{
			mylist.insert_end(i);
		}
		
		mylist.printlist();

		mylist.insert_at_index(2, 21);
		mylist.printlist();
		// mylist.printlist();
		// mylist.insert(2);
		// mylist.printlist();
		// mylist.insert(5);
		// mylist.printlist();
		// mylist.insert(1);
		// mylist.printlist();
		// mylist.insert(69);
		// mylist.printlist();
	}
}

File: py/lab8/ll.py
"""Linked list
Date: 30 June, 2018"""

class node:
	def __init__(self, data = None):
		self.data = data
		self.prev = None
		self.next = None

class ll:
	def __init__(self):
		self.head = node()

	def append(self, data):
		new_node = node(data)
		cur = self.head
		while cur.next != None:
			cur = cur.next
		cur.next = new_node
		new_node.prev = cur

	def insert(self, data, index):
		temp = node(data)
		cur = self.head
		for i in range(index):
			cur = cur.next
		temp.next = cur.next
		cur.next = temp
		
	def length(self):
		length = 0
		cur = self.head
		while cur.next != None:
			cur = cur.next
			length += 1
		return length

	def display(self):
		elem = []
		cur = self.head
		while cur.next != None:
			cur = cur.next
			elem.append(cur.data)
		print (elem)

	def del_ind(self, index):
		if index >= self.length():
			print ("ERROR:\nIndex out of Range")
			return None
		else:
			cur = self.head
			for i in range(0, index):
				cur = cur.next
			prev = cur
			cur = cur.next
			prev.next = cur.next

	def get_data(self, index):
		if index >= self.length():
			print ("ERROR:\nIndex out of Range")
			return None
		else:
			cur = self.head
			for i in range(0, index + 1):
				cur = cur.next
			print ("Value at index", index, "is", cur.data)
	
	def get_index(self, data):
		cur = self.head
		cur = cur.next
		index = 0
		flag = 0
		for i in range(self.length()):
			if cur.data != data:
				cur = cur.next
				index += 1
			elif cur.data == data:
				flag = 1
				print ("First index of value", data, "is", index)
				break
			
		if flag== 0:
			print("Element " + str(data) + " not found")		

def main():
	lis = ll()

	lis.display()
	lis.append(4)
	lis.append(5)
	lis.append(21)
	lis.display()
	print ("list length =", lis.length())
	lis.insert(9, 1)
	lis.display()
	lis.get_index(21)
	lis.get_data(0)
	lis.display()
	lis.get_index(5)
	lis.del_ind(0)
	lis.display()
	lis.get_index(5)

if __name__ == '__main__':
	main()